File: apps/mobile/lib/features/notes/domain/voice.dart
/// Recordings in notes, and notes read aloud ([[Notes]] N7, N10).
library;

/// The extensions an embed may name and still be a recording.
const audioExtensions = {'m4a', 'aac', 'mp3', 'wav', 'ogg', 'opus'};

final _embed = RegExp(r'!\[\[([^\[\]\n]+?)\]\]');

/// Every recording a body embeds, by file name, in order.
///
/// An embed is Obsidian's own syntax, `![[Voice 2026-09-19 14-32.m4a]]`,
/// so the pair of `.md` and `.m4a` opens in Obsidian as it is. Only
/// names with an audio extension count; `![[picture.png]]` is somebody
/// else's business.
List<String> audioEmbedsIn(String body) => [
  for (final match in _embed.allMatches(body))
    if (_isAudio(match.group(1)!.trim())) match.group(1)!.trim(),
];

bool _isAudio(String name) {
  final dot = name.lastIndexOf('.');
  return dot > 0 && audioExtensions.contains(name.substring(dot + 1).toLowerCase());
}

/// The embed line for [fileName].
String audioEmbed(String fileName) => '![[$fileName]]';

/// A recording's name: the moment it started, which is what I remember
/// it by, and unique enough — a second one in the same minute gets a
/// counter.
String voiceFileName(DateTime at, {Set<String> taken = const {}}) {
  String two(int n) => n.toString().padLeft(2, '0');
  final stem =
      'Voice ${at.year}-${two(at.month)}-${two(at.day)} '
      '${two(at.hour)}-${two(at.minute)}';
  var name = '$stem.m4a';
  for (var i = 2; taken.contains(name); i++) {
    name = '$stem ($i).m4a';
  }
  return name;
}

/// A voice note's title, from the moment it was started.
String voiceNoteTitle(DateTime at) =>
    voiceFileName(at).replaceAll(RegExp(r'\.m4a$'), '');

/// A note as it should be heard: the words, without the markdown.
///
/// Headings, emphasis, list markers, quotes, link brackets and code
/// fences are dropped; an embed is dropped whole, because "exclamation
/// bracket bracket Voice two thousand twenty-six" is nobody's idea of
/// reading. What comes back is split into paragraphs, so the player can
/// skip one at a time.
List<String> speechParagraphs(String markdown) {
  final lines = <String>[];
  var inFence = false;
  for (final raw in markdown.split('\n')) {
    final line = raw.trimRight();
    if (line.trimLeft().startsWith('```')) {
      inFence = !inFence;
      continue;
    }
    if (inFence) continue;
    var text = line
        .replaceAll(_embed, '')
        .replaceAll(RegExp(r'^\s{0,3}#{1,6}\s+'), '')
        .replaceAll(RegExp(r'^\s*>\s?'), '')
        .replaceAll(RegExp(r'^\s*([-*+]|\d+[.)])\s+(\[[ xX]\]\s+)?'), '')
        .replaceAllMapped(RegExp(r'\[\[([^\]|]+)(\|([^\]]+))?\]\]'), (m) => m.group(3) ?? m.group(1)!)
        .replaceAllMapped(RegExp(r'\[([^\]]+)\]\([^)]*\)'), (m) => m.group(1)!)
        .replaceAll(RegExp(r'(\*\*|__|\*|_|~~|`|==)'), '')
        .replaceAll(RegExp(r'^\s*([-*_]\s*){3,}$'), '')
        .trim();
    // A table row reads as its cells.
    if (text.startsWith('|')) {
      text = text
          .split('|')
          .map((cell) => cell.trim())
          .where((cell) => cell.isNotEmpty && !RegExp(r'^:?-+:?$').hasMatch(cell))
          .join(', ');
    }
    lines.add(text);
  }
  final paragraphs = <String>[];
  final buffer = StringBuffer();
  for (final line in lines) {
    if (line.isEmpty) {
      if (buffer.isNotEmpty) {
        paragraphs.add(buffer.toString());
        buffer.clear();
      }
      continue;
    }
    if (buffer.isNotEmpty) buffer.write(' ');
    buffer.write(line);
  }
  if (buffer.isNotEmpty) paragraphs.add(buffer.toString());
  return paragraphs;
}

final _arabic = RegExp('[\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF]');

/// Which voice should read [text]: Arabic when most of its letters are
/// Arabic, otherwise [fallback] (the app's language).
String speechLanguageOf(String text, {required String fallback}) {
  final letters = RegExp(r'\p{L}', unicode: true).allMatches(text).length;
  if (letters == 0) return fallback;
  final arabic = _arabic.allMatches(text).length;
  return arabic * 2 >= letters ? 'ar' : fallback;
}

File: apps/web/src/app/components/notes/voice.ts
/**
 * Notes read aloud, dictated and recorded in a browser ([[Notes]] N7,
 * N10), with the phone's rules for what is spoken ported from
 * `voice.dart`.
 *
 * N10 says read aloud and dictation use the device's own engines, and
 * Business rule 13 says nothing new goes out over the network. So both
 * are offered only where the browser can do them on this computer: a
 * voice the browser marks as local, and speech recognition that runs
 * on the device. Where it cannot, the control is not there.
 */

const embed = /!\[\[([^[\]\n]+?)\]\]/g;

/**
 * A note as it should be heard: the words, without the markdown, in
 * paragraphs so the player can skip one at a time (`speechParagraphs`).
 * An embed is dropped whole, and a table row reads as its cells.
 */
export function speechParagraphs(markdown: string): string[] {
  const lines: string[] = [];
  let inFence = false;
  for (const raw of markdown.split('\n')) {
    const line = raw.trimEnd();
    if (line.trimStart().startsWith('```')) {
      inFence = !inFence;
      continue;
    }
    if (inFence) continue;
    let text = line
      .replace(embed, '')
      .replace(/^\s{0,3}#{1,6}\s+/, '')
      .replace(/^\s*>\s?/, '')
      .replace(/^\s*([-*+]|\d+[.)])\s+(\[[ xX]\]\s+)?/, '')
      .replace(/\[\[([^\]|]+)(\|([^\]]+))?\]\]/g, (_, title: string, _pipe: string | undefined, label: string | undefined) => label ?? title)
      .replace(/\[([^\]]+)\]\([^)]*\)/g, '$1')
      .replace(/(\*\*|__|\*|_|~~|`|==)/g, '')
      .replace(/^\s*([-*_]\s*){3,}$/, '')
      .trim();
    if (text.startsWith('|')) {
      text = text
        .split('|')
        .map((cell) => cell.trim())
        .filter((cell) => cell.length > 0 && !/^:?-+:?$/.test(cell))
        .join(', ');
    }
    lines.push(text);
  }
  const paragraphs: string[] = [];
  let buffer = '';
  for (const line of lines) {
    if (!line) {
      if (buffer) paragraphs.push(buffer);
      buffer = '';
      continue;
    }
    buffer = buffer ? `${buffer} ${line}` : line;
  }
  if (buffer) paragraphs.push(buffer);
  return paragraphs;
}

const arabicLetter = /[؀-ۿݐ-ݿࢠ-ࣿ]/g;

/** Arabic when most of the letters are, otherwise [fallback] (`speechLanguageOf`). */
export function speechLanguageOf(text: string, fallback: string): string {
  const letters = text.match(/\p{L}/gu)?.length ?? 0;
  if (letters === 0) return fallback;
  const arabic = text.match(arabicLetter)?.length ?? 0;
  return arabic * 2 >= letters ? 'ar' : fallback;
}

// ------------------------------------------------------------ speaking

/** A voice on this computer for [language], or null: a remote voice would send the note away. */
export function localVoice(language: string): SpeechSynthesisVoice | null {
  if (typeof speechSynthesis === 'undefined') return null;
  const voices = speechSynthesis.getVoices().filter((voice) => voice.localService);
  const base = language.toLowerCase().split('-')[0]!;
  return (
    voices.find((voice) => voice.lang.toLowerCase() === language.toLowerCase()) ??
    voices.find((voice) => voice.lang.toLowerCase().split(/[-_]/)[0] === base) ??
    null
  );
}

/** Whether this browser can speak at all; the voices may still be loading. */
export function canSpeak(): boolean {
  return typeof speechSynthesis !== 'undefined' && typeof SpeechSynthesisUtterance !== 'undefined';
}

// ----------------------------------------------------------- recording

/**
 * The container a recording is made in, and the extension it is named
 * with. The phone records AAC in `.m4a`, and only names with an audio
 * extension it knows count as a recording (N7), so MP4 audio comes
 * first, then Ogg. WebM holds the same Opus audio Ogg would; it takes
 * `.ogg` so the phone keeps it, and players read the bytes, not the name.
 */
export function recordingFormat(): { mimeType: string; extension: string } | null {
  if (typeof MediaRecorder === 'undefined' || typeof navigator === 'undefined' || !navigator.mediaDevices?.getUserMedia) {
    return null;
  }
  const candidates: [string, string][] = [
    ['audio/mp4;codecs=mp4a.40.2', 'm4a'],
    ['audio/mp4', 'm4a'],
    ['audio/ogg;codecs=opus', 'ogg'],
    ['audio/webm;codecs=opus', 'ogg'],
    ['audio/webm', 'ogg'],
  ];
  for (const [mimeType, extension] of candidates) {
    if (MediaRecorder.isTypeSupported(mimeType)) return { mimeType, extension };
  }
  return null;
}

// ----------------------------------------------------------- dictation

interface RecognitionResultList {
  length: number;
  [index: number]: { isFinal: boolean; 0: { transcript: string } };
}

/** The slice of the Web Speech recognition API dictation uses. */
export interface Recognition {
  lang: string;
  continuous: boolean;
  interimResults: boolean;
  processLocally?: boolean;
  onresult: ((event: { resultIndex: number; results: RecognitionResultList }) => void) | null;
  onend: (() => void) | null;
  onerror: ((event: { error: string }) => void) | null;
  start(): void;
  stop(): void;
}

interface RecognitionConstructor {
  new (): Recognition;
  /** Chrome's on-device check; absent where recognition can only use a server. */
  available?: (options: { langs: string[]; processLocally: boolean }) => Promise<string>;
}

function recognitionClass(): RecognitionConstructor | null {
  if (typeof window === 'undefined') return null;
  const scope = window as unknown as { SpeechRecognition?: RecognitionConstructor; webkitSpeechRecognition?: RecognitionConstructor };
  return scope.SpeechRecognition ?? scope.webkitSpeechRecognition ?? null;
}

/**
 * A recogniser that listens on this computer, or null. Recognition that
 * can only work by sending the microphone to the browser's vendor is
 * not offered: that would be a new outbound request (Business rule 13),
 * and N10 promises dictation never leaves the device.
 */
export async function localDictation(language: string): Promise<Recognition | null> {
  const Class = recognitionClass();
  if (!Class || typeof Class.available !== 'function') return null;
  try {
    if ((await Class.available({ langs: [language], processLocally: true })) !== 'available') return null;
  } catch {
    return null;
  }
  const recognition = new Class();
  recognition.lang = language;
  recognition.processLocally = true;
  recognition.continuous = true;
  recognition.interimResults = false;
  return recognition;
}
